velero-plugins/migsa/backup.go

```diff
@@ -0,0 +1,123 @@
+package migsa
+
+import (
+	"encoding/json"
+	"strings"
+
+	v1 "github.com/heptio/velero/pkg/apis/velero/v1"
+	"github.com/heptio/velero/pkg/plugin/velero"
+	apisecurity "github.com/openshift/api/security/v1"
+	security "github.com/openshift/client-go/security/clientset/versioned/typed/security/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+)
+
+// BackupPlugin is a backup item action plugin for Heptio Velero.
+type BackupPlugin struct {
+	Log    logrus.FieldLogger
+	SCCMap map[string]map[string][]apisecurity.SecurityContextConstraints
+}
+
+// AppliesTo returns a velero.ResourceSelector that applies to everything.
+func (p *BackupPlugin) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"serviceaccounts"},
+	}, nil
+}
+
+// This should be moved to clients package in future
+var securityClient *security.SecurityV1Client
+var securityClientError error
+
+// Execute copies local registry images into migration registry
+func (p *BackupPlugin) Execute(item runtime.Unstructured, backup *v1.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+	p.Log.Info("[serviceaccount-backup] Entering ServiceAccount backup plugin")
+
+	serviceAccount := corev1.ServiceAccount{}
+	itemMarshal, _ := json.Marshal(item)
+	json.Unmarshal(itemMarshal, &serviceAccount)
+
+	var additionalItems []velero.ResourceIdentifier
+
+	for _, scc := range p.SCCMap[serviceAccount.Namespace][serviceAccount.Name] {
```

Comment:
If you try to migrate a namespace for which there are no SCCs defined, will `p.SCCMap[serviceAccount.Namespace]` be nil, and if so, will this result in a segfault? We should probably make sure that the map entry for the namespace exists before trying to look up the serviceaccount name in it.